test/test_unistd.rs

```diff
@@ -576,6 +577,39 @@ fn test_symlinkat() {
     );
 }
 
+
+#[test]
+fn test_unlinkat() {
+    let tempdir = tempfile::tempdir().unwrap();
+    let filename = "foo.txt";
+    let dirname = "foo_dir";
+    let filepath = tempdir.path().join(filename);
+    let dirpath = tempdir.path().join(dirname);
+
+    // Create file
+    File::create(&filepath).unwrap();
+
+    // Create dir
+    DirBuilder::new().recursive(true).create(&dirpath).unwrap();
+
+    // Get file descriptor for base directory
+    let dirfd = fcntl::open(tempdir.path(), fcntl::OFlag::empty(), stat::Mode::empty()).unwrap();
+
+    // Attempt unlink dir at relative path without proper flag
+    panic::set_hook(Box::new(|_info| {}));
+    let result = std::panic::catch_unwind(|| unlinkat(Some(dirfd), dirname, UnlinkatFlags::NoRemoveDir).unwrap());
```

Comment:
This is way too complicated.  The right way to test code that ought to panic is with the `#[should_panic()]` attribute.  See sys/test_aio.rs for an example.

Reply:
Hi asomers .. thanks a lot for the feedback. I am definitely new to rust so I appreciate the suggestions. Being fairly new though I did spend time to research the best way to do this and this was the suggestion I came across. Specifically this was suggested over should_panic for more fine grain panic control. For testing this code I wanted the test code to create a folder (and file) and then try to delete the folder first without the proper flag and then try again with. If I use [should_panic()] its my understanding no unexpected panics in the code function would be properly caught hence this suggestion: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/controlling-panics-with-std-panic.html. Any comments on this concern of mine? Looks like below you may be suggesting three different commits which I guess would mitigate my concern of using [should_panic()] where I don't want that coarse grain coverage of the attribute?

Reply:
Yes, if you split the test up, then some tests would have `should_panic` and others wouldn't.  Why do you prefer `std::panic` over the simpler `should_panic`?

Reply:
Hi .. I simply did not want to use an attribute that I realized could hide other unexpected exceptions that should not happen. Using std:panic since was the only suggested way I saw to handle this, plus it was just cutting and pasting two lines of code. Its all unfamiliar syntax for me so it didn't stand out to me as anymore complicated than some of the other lines. Certainly using should_panic should be easier to follow. If I break up this test case as you suggest I am creating more folders/files and using more lines of code but I can then use should_panic. I didn't think to do this initially since I guess I was trying to capture the testing of unlinkat in a single function instead of multiple.

Reply:
Actually, `should_panic` is more specific than you realize.  You can provide an expected string, and the test will fail if it panics with any other message.  See test/sys/test_aio.rs for an example.